### PR TITLE
wasi: add simple wasi test/example

### PR DESCRIPTION
**Pull request by danbev**
This commit adds a very simply wasi example that calls a _start
function on a WebAssembly module that has been compiled with a WASI
compatibly compiler. It depends upon #4 which is included in this PR, but I'll rebase this when that lands.

This is the source c program:
```c
void print() {
  printf("simple-wasi...\n");
}

int main(int argc, char** argv) {
  print();
  return 0;
}
```
And it can be run with wasmtime:
```console
$ wasmtime simple-wasi.wasm
simple-wasi...
```

The test test-wasi-binding.js currently does not print anything as these
functions (WASI::FdPrestatGet etc) are not implemented yet. But I think
it still might be useful to have an example/test that show how this can be
used.

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

File: test/parallel/test-wasi-binding.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+const memory = new WebAssembly.Memory({ initial: 1 });
+const wasi = new WASI({ args: [], env: process.env, memory });
+
+const importObject = {
+  wasi_unstable: wasi.wasiImport
+};
+
+WebAssembly.instantiate(buffer, importObject).then((results) => {
+  assert(results.instance.exports._start);
+  results.instance.exports._start();
```
Comment (cjihrig) [inline test/parallel/test-wasi-binding.js]: I think we'd want to call `WASI.start(results)` here.

File: test/parallel/test-wasi-binding.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+const { WASI } = require('wasi');
+
+const memory = new WebAssembly.Memory({ initial: 1 });
+const wasi = new WASI({ args: [], env: process.env, memory });
+
+const importObject = {
+  wasi_unstable: wasi.wasiImport
+};
+
+WebAssembly.instantiate(buffer, importObject).then((results) => {
```
Comment (cjihrig) [inline test/parallel/test-wasi-binding.js]: Can you wrap this in a `common.mustCall()`.

**Review by cjihrig — APPROVED** [review]
Can you use `test/wasi` and `test/fixtures/wasi` as the directories for the test. I think we'll want to keep the WASI stuff isolated, at least initially, since it requires setting up a special compiler and what not. I also have a branch that I'm getting ready to PR that includes the tests from Gus' PR.

File: test/parallel/test-wasi-binding.js (hunk 1 above)
Comment (danbev) [reply test/parallel/test-wasi-binding.js]: Sure thing, I'll do that.

File: test/parallel/test-wasi-binding.js (hunk 2 above)
Comment (danbev) [reply test/parallel/test-wasi-binding.js]: I'll do that. I'll pushed these changes shortly. Thanks

**Comment by cjihrig** [comment]
Squashed and landed in https://github.com/nodejs/wasi/commit/37a2b525ef8c0dc5a265d47e8af4c903e1eb3ac0. Thanks!

**Pull request closed without merging**